File: python/examples/slimrpc/simple/client_group.py
import argparse
import asyncio
import logging
import time
from datetime import timedelta

import slim_bindings
from examples.constants import (
    NAME_NS,
    NAME_ORG,
    SHARED_SECRET,
    SLIM_ADDR,
)
from examples.slimrpc.simple.types.example_pb2 import ExampleRequest
from examples.slimrpc.simple.types.example_pb2_slimrpc import TestGroupStub

logger = logging.getLogger(__name__)


async def run_multicast_unary(stub: TestGroupStub) -> None:
    logger.info("=== Multicast Unary-Unary ===")
    request = ExampleRequest(example_integer=1, example_string="hello")
    async for context, resp in stub.ExampleUnaryUnary(
        request, timeout=timedelta(seconds=5)
    ):
        logger.info(f"  [{context}] {resp}")


async def run_multicast_unary_stream(stub: TestGroupStub) -> None:
    logger.info("=== Multicast Unary-Stream ===")
    request = ExampleRequest(example_integer=1, example_string="hello")
    async for context, resp in stub.ExampleUnaryStream(
        request, timeout=timedelta(seconds=5)
    ):
        logger.info(f"{time.time()}")
        logger.info(f"  [{context}] {resp}")


async def stream_requests():
    for i in range(3):
        yield ExampleRequest(example_integer=i, example_string=f"item {i}")


async def run_multicast_stream_unary(stub: TestGroupStub) -> None:
    logger.info("=== Multicast Stream-Unary ===")
    async for context, resp in stub.ExampleStreamUnary(
        stream_requests(), timeout=timedelta(seconds=5)
    ):
        logger.info(f"  [{context}] {resp}")


async def run_multicast_stream_stream(stub: TestGroupStub) -> None:
    logger.info("=== Multicast Stream-Stream ===")
    async for context, resp in stub.ExampleStreamStream(
        stream_requests(), timeout=timedelta(seconds=5)
    ):
        logger.info(f"  [{context}] {resp}")


async def amain(server: str, servers_str: str) -> None:
    slim_bindings.uniffi_set_event_loop(asyncio.get_running_loop())  # type: ignore[arg-type]

    tracing_config = slim_bindings.new_tracing_config()
    runtime_config = slim_bindings.new_runtime_config()
    service_config = slim_bindings.new_service_config()

    tracing_config.log_level = "info"

    slim_bindings.initialize_with_configs(
        tracing_config=tracing_config,
        runtime_config=runtime_config,
        service_config=[service_config],
    )

    service = slim_bindings.get_global_service()

    local_name = slim_bindings.Name(NAME_ORG, NAME_NS, "client")

    server_names = [
        slim_bindings.Name(NAME_ORG, NAME_NS, s.strip()) for s in servers_str.split(",")
    ]

    client_config = slim_bindings.new_insecure_client_config(server)
    conn_id = await service.connect_async(client_config)

    local_app = service.create_app_with_secret(local_name, SHARED_SECRET)
    await local_app.subscribe_async(local_name, conn_id)

    # Group channel targeting all server instances
    channel = slim_bindings.Channel.new_group_with_connection(
        local_app, server_names, conn_id
    )

    stub = TestGroupStub(channel)

    print("SLIM_RPC_GROUP_CLIENT_STARTED", flush=True)

    try:
        await run_multicast_unary(stub)
    except slim_bindings.RpcError as e:
        logger.error(f"RPC error in multicast unary-unary: {e}")
        raise

    try:
        await run_multicast_unary_stream(stub)
    except slim_bindings.RpcError as e:
        logger.error(f"RPC error in multicast unary-stream: {e}")
        raise

    try:
        await run_multicast_stream_unary(stub)
    except slim_bindings.RpcError as e:
        logger.error(f"RPC error in multicast stream-unary: {e}")
        raise

    try:
        await run_multicast_stream_stream(stub)
    except slim_bindings.RpcError as e:
        logger.error(f"RPC error in multicast stream-stream: {e}")
        raise

    print("SLIM_RPC_GROUP_CLIENT_DONE", flush=True)

    await asyncio.sleep(1)

    # Close the channel
    await channel.close_async(timeout=None)


def main() -> None:
    parser = argparse.ArgumentParser(description="SlimRPC example group client")
    parser.add_argument(
        "--server",
        default=SLIM_ADDR,
        help="SLIM server endpoint (default: from SLIM_ADDR env var or http://localhost:46357)",
    )
    parser.add_argument(
        "--servers",
        default="server1,server2",
        help="Comma-separated server instance names",
    )
    args = parser.parse_args()

    logging.basicConfig(level=logging.INFO)
    try:
        asyncio.run(amain(args.server, args.servers))
    except KeyboardInterrupt:
        print("Client interrupted by user.")

File: node/examples/slimrpc/simple/client_group.ts
// Copyright AGNTCY Contributors (https://github.com/agntcy)
// SPDX-License-Identifier: Apache-2.0

/**
 * SlimRPC simple example — multicast (group) client.
 *
 * Exercises all four RPC shapes of the generated `TestGroupClient` against a
 * GROUP channel spanning multiple server instances. Each yielded value pairs
 * the responding member's context with its decoded response.
 *
 * Mirrors python/examples/slimrpc/simple/client_group.py and
 * go/examples/slimrpc/simple/cmd/client_group.
 *
 * Run two servers first, e.g.:
 *   npm run server -- --instance server1
 *   npm run server -- --instance server2
 */

import * as slimBindings from '../../../generated/index.js';
import { createAndConnectApp, splitId, logMessage, DEFAULT_SERVER_ENDPOINT, DEFAULT_SHARED_SECRET } from './common.js';
import { create } from '@bufbuild/protobuf';
import { ExampleRequestSchema, type ExampleRequest } from './types/example_pb.js';
import { TestGroupClient } from './types/example_slimrpc.js';

const NAME_ORG = 'agntcy';
const NAME_NS = 'grpc';
const TIMEOUT_MS = 5000;

interface CliArgs {
  local: string;
  servers: string[];
  server: string;
}

function parseArgs(): CliArgs {
  const args = process.argv.slice(2);
  let local = `${NAME_ORG}/${NAME_NS}/client`;
  let servers = ['server1', 'server2'];
  let server = DEFAULT_SERVER_ENDPOINT;

  for (let i = 0; i < args.length; i++) {
    if (args[i] === '--local' && i + 1 < args.length) {
      local = args[++i];
    } else if (args[i] === '--servers' && i + 1 < args.length) {
      servers = args[++i].split(',').map((s) => s.trim()).filter(Boolean);
    } else if (args[i] === '--server' && i + 1 < args.length) {
      server = args[++i];
    } else if (args[i] === '--help' || args[i] === '-h') {
      console.log('Usage: npm run client-group -- [OPTIONS]');
      console.log('');
      console.log('Options:');
      console.log('  --local <id>        Local identity (default: agntcy/grpc/client)');
      console.log('  --servers <names>   Comma-separated server instance names (default: server1,server2)');
      console.log('  --server <address>  SLIM server endpoint (default: http://localhost:46357)');
      console.log('  --help, -h          Show this help message');
      process.exit(0);
    }
  }

  return { local, servers, server };
}

async function* streamRequests(): AsyncGenerator<ExampleRequest> {
  for (let i = 0; i < 3; i++) {
    yield create(ExampleRequestSchema, { exampleString: `item ${i}`, exampleInteger: BigInt(i) });
  }
}

function source(context: { source: { toString(): string } }): string {
  try {
    return String(context.source);
  } catch {
    return '<member>';
  }
}

async function main(): Promise<void> {
  const args = parseArgs();

  console.log('📢 SlimRPC Group Client');
  console.log('=======================');
  console.log(`Local:   ${args.local}`);
  console.log(`Servers: ${args.servers.join(', ')}`);
  console.log(`Server:  ${args.server}`);
  console.log();

  const { app, connId } = await createAndConnectApp(args.local, args.server, DEFAULT_SHARED_SECRET);
  const instance = app.id();

  const serverNames = args.servers.map((s) => splitId(`${NAME_ORG}/${NAME_NS}/${s}`));
  const channel = slimBindings.Channel.newGroupWithConnection(app, serverNames, connId);
  const client = new TestGroupClient(channel);

  console.log('SLIM_RPC_GROUP_CLIENT_STARTED');

  const request = create(ExampleRequestSchema, { exampleString: 'hello', exampleInteger: 1n });

  logMessage(instance, '=== Multicast Unary-Unary ===');
  for await (const { context, response } of client.ExampleUnaryUnary(request, TIMEOUT_MS)) {
    logMessage(instance, `📨 [${source(context)}] "${response.exampleString}" (${response.exampleInteger})`);
  }

  logMessage(instance, '=== Multicast Unary-Stream ===');
  for await (const { context, response } of client.ExampleUnaryStream(request, TIMEOUT_MS)) {
    logMessage(instance, `📨 [${source(context)}] "${response.exampleString}" (${response.exampleInteger})`);
  }

  logMessage(instance, '=== Multicast Stream-Unary ===');
  for await (const { context, response } of client.ExampleStreamUnary(streamRequests(), TIMEOUT_MS)) {
    logMessage(instance, `📨 [${source(context)}] "${response.exampleString}" (${response.exampleInteger})`);
  }

  logMessage(instance, '=== Multicast Stream-Stream ===');
  for await (const { context, response } of client.ExampleStreamStream(streamRequests(), TIMEOUT_MS)) {
    logMessage(instance, `📨 [${source(context)}] "${response.exampleString}" (${response.exampleInteger})`);
  }

  console.log('SLIM_RPC_GROUP_CLIENT_DONE');

  await channel.closeAsync(undefined);
}

process.on('SIGINT', () => {
  console.log('\n\n📋 Received SIGINT — shutting down...');
  process.exit(0);
});

main().catch((error: any) => {
  console.error('Fatal error:', error?.message || String(error));
  process.exit(1);
});
